tpl/tplimpl: Add details shortcode

racehd opened the pull request:
- Add new shortcode to render details HTML element.
- Implement integration tests to check: default state, custom summary, open state, safeHTML sanitization, allowed attributes, and localization of default summary text.
- Update docs to include details shortcode.

Closes #13090

jmooring commented on docs/content/en/content-management/shortcodes.md:
````diff
@@ -94,6 +94,33 @@ Example usage:
+The `details` shortcode allows you to generate a collapsible details HTML element. Example usage:
+
+```text
+    {{</* details summary="Custom Summary Text" */>}}
````
Do not indent example

jmooring commented on docs/content/en/content-management/shortcodes.md:
```diff
@@ -94,6 +94,33 @@ Example usage:
+{{% note %}}
+To override Hugo's embedded `details` shortcode, copy the [source code] to a file with the same name in the layouts/shortcodes directory.
+
+[source code]: {{% eturl details %}}
```
Add the following to data/embedded_template_urls.toml:

```text
'details' = 'shortcodes/details.html'
```

jmooring commented on docs/content/en/content-management/shortcodes.md:
```diff
@@ -94,6 +94,33 @@ Example usage:
+open
+: (`bool`) Optional. Whether to initially display the contents of the details element. Default is `false`.
+
+[...params]
```
```suggestion
```

jmooring replied:
This isn't a named parameter. See below.

jmooring commented on docs/content/en/content-management/shortcodes.md:
```diff
@@ -94,6 +94,33 @@ Example usage:
+: (`bool`) Optional. Whether to initially display the contents of the details element. Default is `false`.
+
+[...params]
+: (`object`) Optional. Additional HTML attributes passed directly to the details element. For instance, can specify `name`, `class`, and other global HTML attributes. 
```
~~~suggestion
You can also pass global HTML attributes when calling the shortcode. For example:

```text
{{</* details summary="Custom Summary Text" class="my-class" */>}}
Showing custom `summary` text.
{{</* /details */>}}
```
~~~

jmooring commented on docs/content/en/content-management/shortcodes.md:
```diff
@@ -94,6 +94,33 @@ Example usage:
+[source code]: {{% eturl details %}}
+{{% /note %}}
+
+The `details` shortcode allows you to generate a collapsible details HTML element. Example usage:
```
```suggestion
Use the `details` shortcode to generate a collapsible details HTML element. For example:
```

jmooring reviewed (commented):
The AssertFileContent method iterates over the values and looks for a match. So it tests if the result contains the value, but it does not test if they are in the given order. I would try to combine these into a single string, possibly using `AssertFileContentEquals` instead.

Also, I think our convention with multiple values is to place the  closing parenthesis on its own line, e.g., 

```text
	b.AssertFileContent("foo",
		"bar",
		"baz",
	)

racehd commented:
> The AssertFileContent method iterates over the values and looks for a match. So it tests if the result contains the value, but it does not test if they are in the given order. I would try to combine these into a single string, possibly using `AssertFileContentEquals` instead.

Thank you for the review! I had noticed AssertFileContent did not check the order. I previously had tried AssertFileContentEquals but it will also check for newlines and whitespace which I thought may make it more prone to triggering false positives in the future. 

The following would pass the checks: 
<details><summary>AssertFileContentEquals</summary>

```
	// Test1: default state (closed by default)
	b.AssertFileContentEquals("public/d1/index.html",
	"\n<details>\n    <summary>Details</summary>\n    <p>Basic example without summary</p>\n</details>\n",
	)
	content1 := b.FileContent("public/d1/index.html")
	c := qt.New(t)
	c.Assert(content1, qt.Not(qt.Contains), "open")

	// Test2: custom summary
	b.AssertFileContentEquals("public/d2/index.html",
	"\n<details>\n    <summary>Custom Summary</summary>\n    <p>Example with custom summary text</p>\n</details>\n",
	)

	// Test3: open state
	b.AssertFileContentEquals("public/d3/index.html",
	"\n<details open>\n    <summary>Test Open State</summary>\n    <p>Example with open state</p>\n</details>\n",
	)

	// Test4: safeHTML sanitization
	b.AssertFileContentEquals("public/d4/index.html",
	"\n<details style=\"color: red;\">\n    <summary>Test Attribute sanitization</summary>\n    <p>Example testing attribute sanitization</p>\n</details>\n",
	)
	content4 := b.FileContent("public/d4/index.html")
	c.Assert(content4, qt.Not(qt.Contains), "onclick")
	c.Assert(content4, qt.Not(qt.Contains), "alert")

	// Test5: class attribute
	b.AssertFileContentEquals("public/d5/index.html",
	"\n<details class=\"custom-class\">\n    <summary>Details</summary>\n    <p>Example with allowed class attribute</p>\n</details>\n",
	)

	// Test6: localization
	b.AssertFileContentEquals("public/es/d6/index.html",
	"\n<details>\n    <summary>Detalles</summary>\n    <p>Localization example without summary</p>\n</details>\n",
	)
```

</details>

Alternatively, I could write a helper function to ignore whitespace and make it a little more robust. Do you have a preference on this?

jmooring commented:
> Do you have a preference on this?

I think what you have is fine.

jmooring commented on tpl/tplimpl/embedded/templates/shortcodes/details.html:
```diff
@@ -0,0 +1,72 @@
+    {{- if $open }} open{{ end }}
+    {{- range $k, $v := .Params }}
+        {{- if not (or (in (slice "open" "summary") $k) (strings.HasPrefix $k "on")) }}
+        {{- printf " %s=%q" $k $v | safeHTMLAttr }}
```
```suggestion
            {{- printf " %s=%q" $k $v | safeHTMLAttr }}
```

jmooring commented on docs/content/en/content-management/shortcodes.md:
```diff
@@ -94,6 +94,38 @@ Example usage:
 Although you can call this shortcode using the `{{</* */>}}` notation, computationally it is more efficient to call it using the `{{%/* */%}}` notation as shown above.
 
+### details
+
```
```suggestion

{{< new-in 0.140.0 >}}

```

jmooring commented on docs/content/en/content-management/shortcodes.md:
````diff
@@ -94,6 +94,38 @@ Example usage:
+{{</* /details */>}}
+```
+
+Additional examples can be found in source code. The `details` shortcode can use the following named arguments:
````
```suggestion
Additional examples can be found in the source code. The `details` shortcode can use the following named arguments:
```

jmooring reviewed (changes requested):
Just three minor changes.

jmooring commented:
Also, please squash commits.

jmooring commented:
Please clean up the commit message and force push. It should look something like this:

```text
- Add new shortcode to render details HTML element.
- Implement integration tests to check: default state, custom summary, open state, safeHTML sanitization, allowed attributes, and localization of default summary text.
- Update docs to include details shortcode.

Closes # 13090
```

You don't need all the messages from the squashed commits.

racehd commented:
Got it, thanks again

jmooring commented:
This looks and works great! I recommend we merge this after the tests go green.

bep commented:
Two comments from me:

We cannot have this:

```handlebars
 {{- printf " %s=%q" $k $v | safeHTMLAttr }}
```

I understand it's ... usefulness, but it would violate Hugo's security policy.

Also:

```
{{- $summary := or (.Get "summary") (T "details") "Details" }}
```

I'm not sure if we're using the pattern above elsewhere, but I think we need to need to namespace the translation keys. I'm not sure what, but `details` seems too generic.

Also, for the future, it would be great if we could wait with the implementation until the issue goes from `Proposal` to `Enhancement`. I understand the desire for speed/momentum, but ...

racehd commented:
Thank you for the feedback. I have made the following modifications:
- Removed `safeHTMLattr`
- Includes namespaced translation key `shortcodes.details`. I found no examples of other shortcode translation keys within Hugo already. 
- Updated to use explicit argument passing for HTML attributes instead of a [,,,params] approach. This felt like it would be less confusing to end-users since `style` would not work without `safeCSS` if a user had multiple style changes (this is also why I did not add `style` as an allowed attribute) 
- Minor formatting of details shortcode. Since attributes are being passed conditionally this inserts a newline to output if each argument is in a new line. To avoid this, i put the allowed attributes on one line to make formatting more predictable for tests. `<details{{- if $open }} open{{ end }}{{- if $name }} name="{{ $name }}"{{- end }}{{- if $class }} class="{{ $class }}"{{- end }}>`

jmooring commented:
@racehd Let's put this on hold for a bit. There's some other work in progress that will affect how we want to handle this. Note that it is desirable to allow some other attributes like `id`, `title`, `data-foo`, etc. Example use of the `data` attribute:

````text
{{< details data-copy=true >}}
```css
body {color: red;}
```
{{< /details >}}
````

Then add a copy button via JS to all `details` element with the `data-copy` attribute.

And in the future, whether added to the embedded template or extended by site and theme authors:

```text
{{< details data-file="examples/something.go" />}}
```
